apps/checkout-app/backend/middlewares.ts

```diff
@@ -0,0 +1,29 @@
+import type { Middleware } from "retes";
+import { Response } from "retes/response";
+import { withSaleorDomainPresent } from "@saleor/app-sdk/middleware";
+import { SALEOR_DOMAIN_HEADER } from "@saleor/app-sdk/const";
+import { getSaleorDomain } from "./utils";
+import { getErrorMessage } from "@/checkout-app/utils/errors";
+
+export const withSaleorDomainMatch: Middleware = (handler) =>
+  withSaleorDomainPresent((request) => {
+    let saleorDomain: string;
+
+    try {
+      saleorDomain = getSaleorDomain();
+    } catch (error) {
+      return Response.InternalServerError({
+        success: false,
+        message: getErrorMessage(error),
+      });
+    }
```

Comment:
I know you had some discussion about `let` here. We can get rid of it by introducing a helper which transforms `Promise<T>` into `Promise<[T, Error]>`:

```ts
type PromiseToTupleResult<T> =
  | [Error, null]
  | [null, T];
const promiseToTuple = async <T>(
  promise: Promise<T>
): Promise<PromiseToTupleResult<T>> => {
  try {
    const result = await promise;
    return [null, result];
  } catch (maybeError) {
    const error =
      maybeError instanceof Error
        ? maybeError
        : new Error(maybeError);
    return [error, null];
  }
};
```

and then:

```ts
const [saleorDomain, saleorDomainError] = await promiseToTuple(getSaleorDomain);
if (saleorDomainError) {
  return Response.InternalServerError({
    success: false,
    message: getErrorMessage(error),
  });
}
// ...
```

* `promiseToTuple` is just a suggestion for a name, definitely not the best one :P